src/platform/graphics/blend-state.js

```diff
@@ -0,0 +1,222 @@
+ *
+ * For the best performance, do not modify blend state after it has been created, but create
+ * multiple blend states and assign them to the material or graphics device as needed.
+ */
+class BlendState {
+    /**
+     * Number bits of which represent the blend state for render target 0.
+     *
+     * @private
+     */
+    target0 = 0;
+
+    /**
+     * Create a new BlendState instance.
+     *
+     * All factor parameters can take the following values:
+     *
+     * - {@link BLENDMODE_ZERO}
+     * - {@link BLENDMODE_ONE}
+     * - {@link BLENDMODE_SRC_COLOR}
+     * - {@link BLENDMODE_ONE_MINUS_SRC_COLOR}
+     * - {@link BLENDMODE_DST_COLOR}
+     * - {@link BLENDMODE_ONE_MINUS_DST_COLOR}
+     * - {@link BLENDMODE_SRC_ALPHA}
+     * - {@link BLENDMODE_SRC_ALPHA_SATURATE}
+     * - {@link BLENDMODE_ONE_MINUS_SRC_ALPHA}
+     * - {@link BLENDMODE_DST_ALPHA}
+     * - {@link BLENDMODE_ONE_MINUS_DST_ALPHA}
+     * - {@link BLENDMODE_CONSTANT}
+     * - {@link BLENDMODE_ONE_MINUS_CONSTANT}
+     *
+     * All op parameters can take the following values:
+     *
+     * - {@link BLENDEQUATION_ADD}
+     * - {@link BLENDEQUATION_SUBTRACT}
+     * - {@link BLENDEQUATION_REVERSE_SUBTRACT}
+     * - {@link BLENDEQUATION_MIN}
+     * - {@link BLENDEQUATION_MAX}
+     *
+     * Note that MIN and MAX operations on WebGL platform require either EXT_blend_minmax or WebGL2
+     * to work (check device.extBlendMinmax).
+     *
+     * @param {boolean} blend - Enables or disables blending. Defaults to false.
+     * @param {number} colorOp - Configures color blending operation. Defaults to
+     * {@link BLENDEQUATION_ADD}.
+     * @param {number} colorSrcFactor - Configures source color blending factor. Defaults to
+     * {@link BLENDMODE_ONE}.
+     * @param {number} colorDstFactor - Configures destination color blending factor. Defaults to
+     * {@link BLENDMODE_ZERO}.
+     * @param {number} alphaOp - Configures alpha blending operation. Defaults to
+     * {@link BLENDEQUATION_ADD}.
+     * @param {number} alphaSrcFactor - Configures source alpha blending factor. Defaults to
+     * {@link BLENDMODE_ONE}.
+     * @param {number} alphaDstFactor - Configures destination alpha blending factor. Defaults to
+     * {@link BLENDMODE_ZERO}.
+     * @param {boolean} redWrite - True to enable writing of the red channel and false otherwise.
+     * Defaults to true.
+     * @param {boolean} greenWrite - True to enable writing of the green channel and false
+     * otherwise. Defaults to true.
+     * @param {boolean} blueWrite - True to enable writing of the blue channel and false otherwise.
+     * Defaults to true.
+     * @param {boolean} alphaWrite - True to enable writing of the alpha channel and false
+     * otherwise. Defaults to true.
+     */
+    constructor(blend = false, colorOp = BLENDEQUATION_ADD, colorSrcFactor = BLENDMODE_ONE, colorDstFactor = BLENDMODE_ZERO,
```

Comment:
I think default blend mode should be SRC_ALPHA, ONE_MINUS_SRC_ALPHA even though blend = false.

Reply:
And probably alpha should default to ADD, ONE, ONE_MINUS_SRC_ALPHA.

Reply:
I preserved existing blend mode, so that's what we had, and I'm not sure there's a reason to change it, as when a blend mode is set, those all are changed anyways.